concourse/steps/release.py

```diff
@@ -23,22 +31,508 @@
+    it atomically, performing the necessary undo actions should an error occur.
+
+    `TransactionalStep`s are provided with access to a shared `TransactionContext` instance
+    to store and retrieve (by step name) information with greater scope than a single step.
+    '''
+    def __init__(
+        self,
+        *steps: TransactionalStep,
+    ):
+        # create context object for this transaction
+        self.context = TransactionContext()
+        # validate type of args and set context
+        for step in steps:
+            if not isinstance(step, TransactionalStep):
+                raise TypeError('Transactions may only contain instances of TransactionalStep')
+            step.set_context(self.context)
+        self._steps = steps
+
+    def validate(self):
+        for step in self._steps:
+            info(f"Validating step '{step.name()}'")
+            step.validate()
+
+    def execute(self):
+        executed_steps = list()
+        for step in self._steps:
+            step_name = step.name()
+            # attempt to execute the step
+            info(f"Applying step '{step_name}'")
+            executed_steps.append(step)
+            try:
+                output = step.apply()
+                self.context.set_step_output(step_name, output)
+            except BaseException as e:
+                warning(f"An error occured while applying step '{step_name}': {e}")
+                traceback.print_exc()
+                # revert the changes attempted, in reverse order
+                self._revert(reversed(executed_steps))
+                # do not execute apply for remaining steps
+                break
+
+    def _revert(self, steps):
+        # attempt to revert each step. Raise an exception if not all reverts succeeded.
+        all_reverted = True
+        for step in steps:
+            step_name = step.name()
+            info(f"Reverting step {step_name}")
+            try:
+                step.revert()
+            except BaseException as e:
+                all_reverted = False
+                warning(f"An error occured while reverting step '{step_name}': {e}")
+                traceback.print_exc()
+        if not all_reverted:
+            raise RuntimeError("Unable to revert all steps.")
+
+
+class ReleaseCommitStep(TransactionalStep):
+    def __init__(
+        self,
+        git_helper: GitHelper,
+        repo_dir: str,
+        release_version: str,
+        repository_version_file_path: str,
+        repository_branch: str,
+        rebase_before_release: bool,
+        release_commit_callback: str=None,
+    ):
+        self.git_helper = not_none(git_helper)
+        self.repository_branch = not_empty(repository_branch)
+        self.release_version = not_empty(release_version)
+        self.rebase_before_release = rebase_before_release
+
+        repo_dir_absolute = pathlib.Path(not_empty(repo_dir)).resolve()
```

Comment:
as already pointed out. please just use `str`. OR replace all usages at least in this module with pathlib.Paths. it is certainly not a good idea to mix styles. and since we need to interface with mako, str will be the more approprate type..